### ISSUE TITLE
<script defer> fires before $(function) in Chrome and FX

### ISSUE THREAD

**Issue by KeithHenry** [jquery/jquery#3271]
In Chrome, FX and IE >=10 `<script defer>` should cause scripts to be background loaded, but then executed in order before `DOMContentLoaded` fires. 

However, when lots of scripts are loaded with `defer` applied the jQuery event fires before all the scripts have run.

You can test this with:

file 1.js:
```javascript
window.test = "fail init"; 
// test result when jQuery ready fires
$(function() {$('#resultJQ').html("jQuery result: " + window.test);});

// test result when document is ready
document.addEventListener("DOMContentLoaded", function() {$('#resultDOM').html("DOM result: " + window.test);});
```

file 2-49.js:
```javascript
window.test = "fail " + n; 
```

file 50.js:
```javascript
window.test = "pass" 
```

page.html:
```html
<!DOCTYPE html>
<html>
<head>
    <script type="text/javascript" src="jquery.js" defer></script>
    <script type="text/javascript" src="1.js" defer></script>
    ...
    <script type="text/javascript" src="n.js" defer></script>
    ...
    <script type="text/javascript" src="50.js" defer></script>
</head>
<body>
    <div id="resultJQ">waiting...</div>
    <div id="resultDOM">waiting...</div>
<body>
</html>
```

Both tests pass when files load fast enough, but sufficient delays (>50ms each) in the `*.js` files cause the jQuery test to fail in latest Chrome and FX. It doesn't fail consistently, but the jQuery ready always fires before the document one.

This bug is extant in the latest 2.x and 3.x but doesn't seem to occur in 1.x

This appears to be due to this check:

```javascript
// Catch cases where $(document).ready() is called
// after the browser event has already occurred.
// Support: IE <=9 - 10 only
// Older IE sometimes signals "interactive" too soon
if ( document.readyState === "complete" ||
	( document.readyState !== "loading" && !document.documentElement.doScroll ) ) {

	// Handle it asynchronously to allow scripts the opportunity to delay ready
	window.setTimeout( jQuery.ready );
}
```

When debugging it appears that `document.readyState === 'interactive'` and `!document.documentElement.doScroll` is true, so `jQuery.ready` fires with the minimum delay (4ms). Breaking on `window.setTimeout` halts execution while some of the numbered js files are still `pending` in the Network tab.

I don't know whether this is a recent change in behavior, but Chrome and Firefox appear to report `document.readyState === 'interactive'` as soon as the body is present, but before deferred scripts are evaluated.

This isn't a problem in IE < 10 as it doesn't support `defer` (it sort of does, but fires scripts out of order and so would fail this test anyway).

**Comment by dmethvin** [comment]
What a refreshing change! Our recent complaints have been about `.ready()` firing too late, and here's one about it firing too soon. :smile: 

There's a long history with us trying to use "interactive" and it looks like we may just need to revert it and put a comment there that says `// NO REALLY YOU CAN'T`. See dabd5ba96c05279b3ffb052db5b8d17f75996694, [gh-2100](https://github.com/jquery/jquery/issues/2100), and [trac-10067](https://bugs.jquery.com/ticket/10067).

**Comment by KeithHenry** [comment]
I think having events fire on `interactive` makes sense, especially if you're working to reduce the time to first meaningful interaction, but it can't be the same as ready - there are two events:

* Do something as soon as the user can interact with the page.
* Do something once all the javascript libraries that need to be loaded have been.

Both are needed, but they're not the same thing.

**Comment by dmethvin** [comment]
@KeithHenry I agree both can be useful, but `$(fn)` has to have its semantics defined. We either need to ensure `fn` doesn't run before `defer` scripts load/execute or say that we don't support loading jQuery itself using the `defer` attribute.

**Comment by KeithHenry** [comment]
Exactly.

I'd argue that `$(document).ready(function)` should mean fully ready, and `$(document).interactive(function)` would be new functionality.

As long as it's clear and explicit what event jQuery is firing on then that should resolve this bug.

**Comment by timmywil** [comment]
The problem is there doesn't seem to be a way to get in before the event loop starts, but after scripts have loaded and executed. You either have to wait until the ready state is `complete`, which is potentially many event loops after `DOMContentLoaded` or stick with `interactive`, which seems to be the right move in most cases.  In other words, dropping `interactive` makes ready just like `load`, and the only advantage is that jQuery can be loaded with the `defer` attribute. If all the scripts were moved to the end of the body, you'd get the same advantage of parsing all the DOM before parsing/executing scripts without the need to drop `interactive`.

I wish there was a web API to actually check DOMContentLoaded, but the ideal would be something that behaves like this.

```html
<script>
document.addEventListener('DOMContentLoaded', function() {
  window.DOMLoaded = true;
});
</script>
<script type="text/javascript" src="jquery.js" defer></script>
<script type="text/javascript" src="1.js" defer></script>
...
<script type="text/javascript" src="10.js" defer></script>
```

Then we (or the user) could check `window.DOMLoaded`. Neither `complete` nor `interactive` really cuts it.

All that to say, a perfect `.ready` is impossible. This is a choice between `interactive` or `complete`. It pains me to say it, but I think we have to drop `interactive`. It would be worse to document that jQuery can't be loaded with the `defer` attribute.

**Comment by neoncube2** [comment]
Do you think something like this would work?

```
if ( document.readyState === "complete") {
	window.setTimeout( jQuery.ready );
}
else if( document.readyState === "interactive" ) {
    document.onreadystatechange = function () {
      if ( document.readyState == "complete" ) {
          window.setTimeout( jQuery.ready );
      }
  }
}
else {
    document.addEventListener( "DOMContentLoaded", completed );
  
    window.addEventListener( "load", completed );
}
```

The idea is that if `ready()` is called when `readyState` is `completed`, then the ready handler would fire pretty much immediately, but if `readyState` was `interactive`, then it would wait until `readyState` was `completed` to fire. If `readyState` was neither `completed` or `interactive`, then the `DOMContentLoaded` event would be used.

I think this should preserve the desired behavior except for in some cases where `defer` is used, in which case `ready()` would be the same as `load()`, which seems like it might be okay.

**Comment by mgol** [comment]
@timmywil If a conclusion is to revert the workaround perhaps we could land it in 3.2.1 instead of waiting for 3.3.0? The patch will be simple, the main problem is testing that thing.

**Comment by gkatsanos** [comment]
TL/DR is "don't load jQuery with defer" ? I struggled hours with a bug before I ended up here :/

**Comment by neoncube2** [comment]
@gkatsanos Yes, I believe that's the current state of things.

**Comment by neoncube2** [comment]
I've done some testing, and it appears that replacing the code that tests for ready state with the following code fixes the issue for me, at least on a Chromium browser.

```
// Catch cases where $(document).ready() is called
// after the browser event has already occurred.
if ( document.readyState === "complete" ) {

	// Handle it asynchronously to allow scripts the opportunity to delay ready
	window.setTimeout( jQuery.ready );
} else if ( document.readyState === "interactive" ) {
    document.onreadystatechange = function () {
      if ( document.readyState == "complete" ) {
          window.setTimeout( jQuery.ready );
      }
  }
} else {

	// Use the handy event callback
	document.addEventListener( "DOMContentLoaded", completed );

	// A fallback to window.onload, that will always work
	window.addEventListener( "load", completed );
}
```

If the document is in the "interactive" state, then the code waits for the document to be in the "complete" state and then runs the ready handler. This means that when jQuery is loaded with "defer", the ready handler is fired only when the page has completely finished loading and any other deferred scripts have run, but when jQuery is not loaded with defer, it tries to use DOMContentLoaded() or immediately fires the ready handler.

The idea is that DOMContentLoaded is still used except for when jQuery is loaded with "defer", in which case it waits for the whole document to be completely loaded before the ready handler is fired. That isn't ideal, but I think it's better than always waiting for the document to be completely loaded before the ready handler is run.

Hopefully this is helpful and I'm not missing something. If anyone could test this out, that would be great.

**Comment by timmywil** [comment]
@neoncube2 That sounds like a reasonable compromise.

**Comment by KeithHenry** [comment]
@neoncube2 Won't that just break it for the users that expect `ready` to fire early? I don't think the issue here is that it's hard to fix, I think it's that an unknown number of existing jQuery components out there expect `ready` to fire early and would break if that behaviour changes to fire when the page is _actually_ ready.

Fixing this would potentially reopen https://github.com/jquery/jquery/issues/2100

**Comment by timmywil** [comment]
@KeithHenry Yea, I just re-read the code. Looks like that's just waiting for load. I was thinking there could be a compromise that wouldn't change behavior for those not using `defer`.

**Comment by neoncube2** [comment]
@KeithHenry and @timmywil 

Ah, I think you may be right.

Is what I've outlined below the behaviour that we're going for, then?


**For when jQuery is loaded with `defer`:**

- If `readyState` is `interactive`, run the ready handler, but only after all other scripts that were loaded with `defer` have run.
- If `readystate` is `complete`, immediately run the ready handler
- Otherwise, run the ready handler on `DOMContentLoaded()`

**For when jQuery isn't loaded with `defer`:**

- If `readyState` is `interactive` or `complete`, run the event handler immediately.
- Otherwise, run the ready handler on `DOMContentLoaded()`

Is that right?

Also, @timmywil mentioned that if we could tell if DOMContentLoaded() had already been fired, that might be helpful, and I think that might be possible using [PerformanceTiming.domContentLoadedEventStart](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming/domContentLoadedEventStart), although that might not be the direction you want to go.

**Comment by KeithHenry** [comment]
@neoncube2 not really - jQuery is already good at firing ready if called after the DOM events - that isn't the issue.

The problem is that that `ready` fires as soon as the page is interactive, and an unknown number of projects that depend on jQuery rely on that behaviour.

To work when `defer` is set jQuery needs to delay `ready` until `document.readyState === 'complete'`, and _never_ fire when `document.readyState === 'interactive'` but doing so will be a breaking change for these unknown projects. There may even be projects out there that use `defer` and rely on the current broken behaviour.

I suppose you could pass some kind of `defer-mode` to jQuery in order to change the behaviour, but then you're either adding to an already fairly large library or splitting the code to deliver an alternate variant, and adding complexity and technical debt either way.

**Comment by alex-jitbit** [comment]
I can confirm that **even when jQuery is loaded synchronously** (without `defer`) but the page has many `defer` scripts and some of those have some latency - the `ready` event fires before some of the scripts have been loaded.

We run a relatively big app and our sentry error tracker quickly fills with confirming errors especially from users on mobile.

**Comment by mgol** [comment]
@alex-jitbit Thanks for the info. Is the large number of deferred scripts needed to hit that issue or is a single bug long-loading one enough to hit it?

**Comment by alex-jitbit** [comment]
@mgol It used to a be a single load-loading script (I added an artificial delay server-side), but I can't replicate it this morning! Tried in both Chrome, Firefox and even IE.

Terribly sorry for the false alarm. I will delete my comments later to keep the issue clean.

**Comment by shmdhussain** [comment]
what is the current status of this Issue ,

This issue is affecting our application which uses angularjs(1.x), i.e jquery ready event causes the angularjs bootstrap,  before all the deferred dependency scripts loaded for angularjs.

jQuery version: 3.4.0

**Comment by dmethvin** [comment]
@shmdhussain The ticket has all that information. Open, assigned to timmywil, scheduled for 4.0 since it is a breaking change for people depending on the current behavior.

**Comment by dperini** [comment]
@dmethvin @timmywil 
I leave here this slightly old patch I had prepared for version 3.2.0 as an attempt to solve this issue.
It needs review and testing but lately my free time is so short, however it looks like these few lines are enough to make the test in this issue to pass and seems it could work also in other circumstances.
Let me know what other requirements exists out of this test or provide new tests.

--- jquery-3.2.0.js.orig	2017-03-16 22:30:24.000000000 +0100
+++ jquery-3.2.0.js	2020-02-25 21:33:13.805507664 +0100
@@ -3046,6 +3046,14 @@
 		// HANDLE: $(function)
 		// Shortcut for document ready
 		} else if ( jQuery.isFunction( selector ) ) {
+			if (root.ready !== undefined) {
+				document.addEventListener('load', function(e) {
+					var s = document.scripts;
+					if (e.target === s[s.length - 1]) {
+						root.ready( selector );
+					}
+				}, true);
+			}
 			return root.ready !== undefined ?
 				root.ready( selector ) :

**Comment by dperini** [comment]
@KeithHenry 
as you are the original author of this issue, I would like you to tell if the above patch does complpetely fix your issue?
And, in case it works, do you see other issues with the suggested patch or other "defer" tests that it doesn't pass ?

I have tested it myself several times and this change seemed to fully pass your test.
However I haven't get any feedback on it, so I suspect I am missing some insight.
Thank you for your help.

**Comment by KeithHenry** [comment]
@dperini The problem was never that my issue was hard to fix - just removing the case where `document.readyState === 'interactive'` could fire `ready` resolves this (though my problem was in an application that migrated away from jQuery because of this issue, at a company I left two months after raising this 4 years ago, so difficult to retest).

I think the problem is that while `ready` should _never_ fire before `defer` scripts, it does now and lots of stuff out there depends on that behaviour. Fixing this issue will break (or rollback) other resolved issues like https://github.com/jquery/jquery/issues/2100

So yes, I think it would fix my issue (although a link to the whole commit or PR would be a lot easier to read than a copy-paste from the change, I can't see which lines you've removed), but the PR with that change may well break other components that rely on jQuery firing before `defer` scripts have finished loading. I think it would have to be a breaking change between major versions.

**Comment by dperini** [comment]
@KeithHenry 
there were no deletion and absolutely no other changes to version 3.2.0, with the exception of the 8 lines added above.
And, since nothing else has changed, there is a chance to use a flag to opt-in this behaviour if that's what is needed at times.

**Comment by KeithHenry** [comment]
@dperini Oh, then that doesn't fix this. As long as `ready` fires for `document.readyState !== "loading"` this behaviour will persist whenever the deferred script takes longer to load the interactive state in the page. The problem is not that `ready` doesn't fire on load, it's that it fires _before_ load.

Really, `interactive` and `ready` should be different events, but we have a bug that `ready` and `$(function)` fire on `interactive`.

**Comment by dperini** [comment]
Don't take me wrong, or think I am presumptuous, but I have to ask if you fully understand what that "addEventListener" does ?
Those 8/10 lines has nothing to  do with the page load event nor with the page "readyState" (interactive, loading, complete).
The difference is so subtle (capture flag true/false) that I have to make sure we are talking about the same behaviour.

I have almost no doubt you know what you say just ensuring that the subtle difference doesn't trick you to believe it has to do with the page "load" event. It instead has to do with the page resources "load" events, in this case with each script "load" event.

I say this because I see you always refer to "page" states and the 4 values assumed by the document "readyState" property.
Again, sorry if I am wrong about that.

**Comment by KeithHenry** [comment]
@dperini No, I get you. Your code (in isolation) would work just fine.

The problem is this line, which you aren't removing: https://github.com/jquery/jquery/blob/25407dcaf5b103ea1055225be502127a34001adc/src/core/ready.js#L67-L76

That causes `ready` to fire too early (though not consistently, it depends on how long the deferred content takes to load and whether `setTimeout` fires before or after them). Here is a simple fix for that:

```js
if ( document.readyState === "complete" ) {
        // #3271 don't fire ready early
	// ( document.readyState !== "loading" && !document.documentElement.doScroll ) 

	// Handle it asynchronously to allow scripts the opportunity to delay ready
	window.setTimeout( jQuery.ready );
} ...
```

Now the early `ready` issue is fixed. The problem is that lots of jQuery dependents rely on `ready` firing early - this would be a _breaking change_ for them. This issue can't be fixed in 3.x jQuery without creating a lot more issues. 

Still, you could give it a go - make a change to `src/core/ready.js` (not the built `jquery-3.2.0.js` output) in your own branch and then link to this issue on the PR. If it gets merged in this issue will be closed and you'll become a contributor, if not you'll get a lot more discussion about your specific change.

**Comment by adjenks** [comment]
In case this helps anyone else. I found that Chrome has load sequence bugs for script tags with defer attributes in xhtml pages:
    https://bugs.chromium.org/p/chromium/issues/detail?id=874749
    https://bugs.chromium.org/p/chromium/issues/detail?id=611136
Basically, if your page is returning xhtml in the content-type header, the page will use the xhtml engine and the scripts may load in the wrong order if they have the defer attribute.
Very few people use xhtml, but this might help someone...

**Comment by ajm19826** [comment]
Soon 10 years, any news?
